### ISSUE TITLE
Incorporate task_form checkboxes and select list value to form handling

### ISSUE THREAD

**Issue by bsaiken** [FWAJL/FieldWorkAssistantMVC#437]

**Comment by FWAJL** [comment]
@bsaiken 

Please describe what you mean by form handling? Do you mean store the data in the DB?

Thank you.

**Comment by bsaiken** [comment]
Yes.  The checkbox values need to be stored so the next time the user 
opens the task the proper tabs are display and the trigger values are 
recorded.  Also, if it is a recurring trigger, the frequency value needs 
to be stored.

Brian

**Comment by damirius** [comment]
@bsaiken @FWAJL 

Since this is the next thing after field form listing, I wanted to ask few questions.
I see there are 3 fields, `task_trigger_cal`, `task_trigger_pm` and `task_trigger_ext`. By input name attributes in current task form I see that "Manual" is `task_trigger_pm`, "External" is `task_trigger_ext`. Since Manual is just a bool flag, we can store it as a tinyint (0/1) and `task_trigger_pm` is already set up like that, but  "External" isn't, it's actually varchar with 50 char length. Do we need to store something else there?
Also I don't know what's `task_trigger_cal` for and if it's used for storing selected recurring period from the first checkbox.

Other part of the form "Data collection and services" isn't part of this issue?

**Comment by bsaiken** [comment]
@damirius 
You will need a bool value in the database no matter what to record whether or not the boxes are checked.  That way, if the user navigates away from the page, the checked box value is not lost.

I will explain the function of each trigger and you can decide what is required to store the information.

Once a Task is prepared, we need to "start" the task.  Tasks get started one of three ways.  This is actually true of all tasks, if you think about it.

Manually.  This means somebody (the PM) says "Time to start".  Nothing happens unless the PM says so.

Recurring.  Some tasks repeat at a set frequency.  Let's say you need to water the lawn once a week or once a month.  You don't wait for the PM, you water the lawn every week.

External.  Some tasks only begin when there is an external trigger.  A fire alarm is a good example.  The fire fighters do nothing unless there is an alarm, or external trigger, then they go to work.

Some tasks depend on more than one trigger.  
Let's say you are supposed to water the lawn weekly, but only if your PM says so.  If your PM says, "Go water the lawn", but you watered it yesterday and you only need to water weekly, you don't go because the _trigger_cal value is not met.  
Say it has been a week, and your PM says "water the lawn", but it is raining.  The external trigger (rain) means you don't water the lawn.  In this case all three triggers must be met to start the task.

The value stored at `ask_trigger_pm`t is a bool - this just answers the question: does the PM need to start the task?  If yes, there will be a button in the "Active Task UI" for the PM to start the task.   

The value stored at `task_trigger_cal` is a bool.  This just answers the question: does this task repeat?  You also need to know the frequency (weekly, monthly, etc.).  Technically you could get by without the bool because if the the `freq_list_box` value is NULL, or not one of the list items, then `task_trigger_cal` must be "0".  Maybe we do need the `task_trigger_cal` value in case the user checks the box, but forgets to choose a value from the dropdown list.  

The value stored at `task_trigger_ext` is a bool.  This just answers the question: is this task affect by an external trigger?  All you need to do is the checkbox.  I am not going to implement external triggers at this stage.

**Comment by damirius** [comment]
@bsaiken

OK, to sum it up we need to switch all of them to bool, or rather tinyint values in the db.
Now for `task_trigger_cal` which should be used for recurring tasks we can leave it like nullable varchar, if it is `NULL` it isn't selected, if it is empty string it is selected but user forgot to select the specific frequency and if it have some other value like "Daily", "Weekly", etc. it is selected and frequency is chosen. Or we can just turn it to tinyint and add new field in the db table, let's say, `task_trigger_cal_value` which will contain actual frequency.
From the top of my head only difference between this two solutions is that if we use two fields as the storage we could preselect some frequency from the dropdown menu.
Let me give you example of this.
I check "Recurring" checkbox and select "Daily" from the frequency dropdown. We will store `1` in `task_trigger_cal` and "Daily" in `task_trigger_cal_value`. I change my mind later and uncheck "Recurring" checkbox. We store `0` in `task_trigger_cal`. OK, so I messed up again and I want to activate "Recurring" checkbox again, dropdown pops up and "Daily" frequency is already selected because it is stored already in the `task_trigger_cal_value`.
If we would store it in the same field we couldn't do that because we would loose that value when storing `NULL` in the `task_trigger_cal`.
It's rather a small thing and I have no idea if it will ever come to that and if it matters, but I just wanted to point it out.
In any case we can do this however you like.

Thanks!

**Comment by bsaiken** [comment]
@damirius 
I am working on the SOW for the Active Task UI, and I realize we will need the bool values for `task_trigger_cal`, `task_trigger_pm`, and `task_trigger_ext.  There is going to be some logic that depends on whether or not task_trigger_x = 1.

**Comment by damirius** [comment]
@bsaiken

Cool. We can go with 2 fields then for "recurring", `task_trigger_cal` and `task_trigger_cal_value`

**Comment by damirius** [comment]
@bsaiken

Added task trigger saving from task form. Don't forget to update your db! You can find the update script in Database\Scripts\Updates\2015-03-16.sql.

**Comment by bsaiken** [comment]
@damirius 

Looks good.  I meant for class="task_boxes" to be included as well.

Brian

**Comment by bsaiken** [comment]
@damirius 

I meant the class="tab_cbox" items, that control which tab is visible.

**Comment by damirius** [comment]
@bsaiken

I've added those 4 checkboxes in the task table. Update your db with this script Database\Scripts\Updates\2015-03-17.sql. I haven't linked them with the tabs since I wanted to ask you few questions first.
First off, those were already linked to what I assume were placeholder tabs. "Field matrix", "Lab matrix", "Forms" and "Chain of custody".
I guess I need to link them like this:
"Task requires completion of a field form" (`task_req_form`) => Forms (new tab for the next issue),
"Analyzing field samples" (`task_req_field_analyte`) => Field Analytes,
"Submitting samples to a laboratory" (`task_req_lab_analyte`) => Lab Analytes,
"Outside Resources required" (`task_req_services`) => Services

If my guess is wrong please do tell me. Checkbox state is currently saved but tabs will not show/hide based on it, I'll see that I link them when we get this sorted out. I'll move onto the next issue while I am waiting for your confirmation.

Thanks!

**Comment by bsaiken** [comment]
@damirius 
Linking checkboxes to tabs is already done in tabs.js.  All you need to do is save the checkbox value to the database.

**Comment by bsaiken** [comment]
@damirius 
This is ready for code review then?

**Comment by damirius** [comment]
@bsaiken

As I said checkboxes are linked to "Field matrix", "Lab matrix", "Forms" and "Chain of custody". Those tabs are just placeholder tabs and they aren't done. All tabs added are called differently and aren't linked to those checkboxes. "Lab analytes", "Field analytes", "Services" are added not so long ago, but they aren't linked to the checkboxes. So the question still stands, do we need to link them to the proper tabs?

**Comment by damirius** [comment]
@bsaiken

Let me also give you example of this, go on http://dev.fieldworkmanager.com. Login and choose some project with the tasks, go to task->edit and check some of those 4 checkboxes under "Data collection and services".
Is this the functionality that we want? 
"Task requires completion of a field form", shows "Forms" tab.
"Analyzing field samples", shows "Forms" and "Field Matrix" tab.
"Submitting samples to a laboratory", shows "Lab Matrix" and "Chain of custody" tab.
"Outside Resources required" isn't showing anything.

Plus first two checkboxes "Task requires completion of a field form" and "Analyzing field samples" are sharing "Forms" tab, checking both of them and then unchecking one will hide the "Forms" tab even when the other is active.
Also to get back to my initial question, is "Lab Matrix" = "Lab analytes" and "Field matrix" = "Field analytes"? I can remove the comment from that part of the code on my branch in the next commit if you wish, just wanted to clear some things out since I don't think this is intended purpose of it.

**Comment by bsaiken** [comment]
@damirius 
The short answer is: all I wanted you to do was incorporate the checkboxes, so that the value is stored in the database.  Don't overthink the job.  I believe you are finished.  I will do the function review shortly.

The tabs.js controlling tabs behaves as I intend at this stage.  I will handle the js associated with the checkboxes.  The behavior may seem odd to you, but it is what I want.  We have commented out some of the js for development purposes - sometimes it is easier to work on a tab if it is always visible.

There are 10 tabs on the Task UI.  They are all different.  "Lab Matrix" != "Lab analytes" and "Field matrix" != "Field analytes".

Once the tabs are all functional, it will be easier for me to explain their purposes.

Brian

**Comment by damirius** [comment]
@bsaiken

Thank you for your explanation.

**Comment by bsaiken** [comment]
@damirius 
I see what you did now.  You added a new tab, `data-form-id="task_forms"` with `id="tab8"`.  This is incorrect.

You need to give this tab `id="tab2"` and replace the existing tab2 (`data-form-id="task_insp_info"').  The order the tabs are displayed matters, so it cannot be located at the end of the list.  

If you do this, the js tab behavior (tabs.js lines 47 - 98) will be correct.

Lines 26 - 45 of tabs.js may need to be modified because the correct tabs display only if the user toggles the checkboxes while in the Task UI.  The correct tabs are not corretly displayed (depending on checkbox status) when the Task is first loaded.

We can fix this later - one Issue at a time...

**Comment by damirius** [comment]
@bsaiken

OK, Form tab is now tab2. Also I've removed commented lines from the checkbox show/hide functions which I've added in previous commit. Also when `loadEditForm` function is ran in task_manager.js and values are fetched from the db "checked" tabs should be displayed. Of course changing tabs will hide them again, but as you said one issue at a time.

**Comment by bsaiken** [comment]
@damir @FWAJL 
I am pushing this to code review, but Damir you need to comment out lines 48, 53, 57, 67.  Otherwise the tab can disappear after it is selected, which is bizarre. 

I will write a new Issue explaining the tab behavior I want...

**Comment by bsaiken** [comment]
@damirius 
Damir you need to comment out lines 48, 53, 57, 67 in `tabs.js`. Otherwise the tab can disappear after it is selected, which is bizarre.  See above comment, where your name got mysteriously truncated.

**Comment by FWAJL** [comment]
@damirius 

I will wait you finish what @bsaiken is asking. Set the status back to code review once it is done.

Thank you.

**Comment by damirius** [comment]
@FWAJL 

Done
